File: Problems/Algorithms/1770. Maximum Score from Performing Multiplication Operations/max_score_multipliers.py
#!/usr/bin/env python3

from functools import lru_cache
from typing import List


class Solution:
    def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
        n, m = len(nums), len(multipliers)
        if m == 1:
            return max(nums[0] * multipliers[0], nums[-1] * multipliers[0])

        @lru_cache(1000)
        def dp(i, j, k):
            if k == m:
                return 0
            return max(nums[i] * multipliers[k] + dp(i + 1, j, k + 1),
                       nums[j] * multipliers[k] + dp(i, j - 1, k + 1))

        return dp(0, n - 1, 0)

File: Problems/Algorithms/1770. Maximum Score from Performing Multiplication Operations/MaxScoreMultipliers.kt
class Solution {
    private lateinit var memo: MutableMap<Int, MutableMap<Int, Int>>

    fun maximumScore(nums: IntArray, multipliers: IntArray): Int {
        val n = nums.size
        val m = multipliers.size
        memo = mutableMapOf()

        return dfs(nums, multipliers, 0, 0)
    }

    private fun dfs(nums: IntArray, multipliers: IntArray, i: Int, j: Int): Int {
        if (j == multipliers.size) return 0
        if (memo[i]?.get(j) != null) return memo[i]!![j]!!

        val n = nums.size
        val m = multipliers.size
        val left = nums[i] * multipliers[j] + dfs(nums, multipliers, i+1, j+1)
        val right = nums[n-1-(j-i)] * multipliers[j] + dfs(nums, multipliers, i, j+1)
        memo[i] = memo.getOrDefault(i, mutableMapOf())
        memo[i]!![j] = maxOf(left, right)
        return memo[i]!![j]!!
    }
}
